shared/domains/chelonia/GIMessage.js

```diff
@@ -46,44 +56,57 @@ export class GIMessage {
     op: GIOp,
     signatureFn?: Function = defaultSignatureFn
   ): this {
-    const message = {
+    const head = {
       version: '1.0.0',
       previousHEAD,
       contractID,
-      op,
-      // the nonce makes it difficult to predict message contents
-      // and makes it easier to prevent conflicts during development
-      nonce: Math.random()
+      op: op[0]
     }
+    console.log('createV1_0', { op, head })
+    const message = op[1]
     // NOTE: the JSON strings generated here must be preserved forever.
     //       do not ever regenerate this message using the contructor.
     //       instead store it using serialize() and restore it using
     //       deserialize().
+    const headJSON = JSON.stringify(head)
     const messageJSON = JSON.stringify(message)
+    const signedPayload = blake32Hash(`${blake32Hash(headJSON)}${blake32Hash(messageJSON)}`)
```

Comment:
Why are we hashing 3 times?

Reply:
@corrideat Hmm. Sorry I still don't understand. Why is it a problem that the boundary isn't well defined, what attack are you referencing?

Reply:
Let's say you have two messages.

```
var msg1 = 'foo'
var msg2 = 'bar'

var hash_a =H(msg1 + msg2)

// Now, let's say you have two different messages
var msg1 = 'foob'
var msg2 = 'ar'

var hash_b =H(msg1 + msg2)

// This works
assert(hash_a === hash_b)
```

Like mentioned, this isn't so easy to do with JSON, because AFAIK it doesn't quite work to extend the previous message with parts of the new message. There could be some clever trick I'm not thinking of, but otherwise you end up with invalid JSON (the attack still 'works' but it'll fail once the JSON is parsed).

Since computing a hash isn't a particularly expensive operation, I'd rather have the cryptography be as strong as it can be in general, as if the message encoding were to change, the attack above might work.

I'm no cryptographer, but another thing I've heard when it comes to security proofs is that having things be fixed length sometimes makes proofs easier.

Reply:
I still don't see what the attack is... Both the headJSON and the messageJSON come from the same data, and we just need a hash of that data, and a signature on top of that hash. If no attack exists I see no reason to have 2 hashes inside of the hash... would recommend removing to avoid unnecessary steps.

Reply:
The attack is that the content of the fields can be manipulated to give the desired hash in certain situations and thus forge signatures. It's much simpler to neutralise the attack by hashing the contents separately instead of analysing every possible scenario to exclude the attack.

Now, there's an unnecessary hash, which is the outer hash. That was mainly to save a few bytes of memory since the hash to be used for signatures is stored in the object.

Reply:
OK, I think I get what you're saying ... it's because of the head split, right, I forgot. Resolved!